packages/ckeditor5-image/src/image/utils.js

```diff
@@ -259,6 +283,26 @@ export function getImageTypeMatcher( matchImageType, editor ) {
 	};
 }
 
+/**
+ * Considering the current model selection, it returns the name of the model image element
+ * (`'image'` or `'imageInline'`) that will make most sense from the UX perspective if a new
+ * image was inserted (also: uploaded, dropped, pasted) at that selection.
+ *
+ * The assumption is that inserting images into empty blocks or on other block widgets should
+ * produce block images. Inline images should be inserted in other cases, e.g. in paragraphs
+ * that already contain some text.
+ *
+ * @param {module:core/editor/editor~Editor} editor
+ * @param {module:engine/model/selection~Selection|module:engine/model/documentselection~DocumentSelection} selection
+ * @returns {'image'|'imageInline'}
+ */
+export function determineImageTypeForInsertionAtSelection( editor, selection ) {
+	const schema = editor.model.schema;
```

Comment:
Also, this is missing the `editor.config.get( 'image.insert.type' )` to allow forcing default image type on insertion.

Reply:
> Also, this is missing the `editor.config.get( 'image.insert.type' )` to allow forcing default image type on insertion.

I'm not sure whether this tool should respect this configuration. 

When used in `insertImage()`, some other logic handles this configuration and that's OK. 

If we're going to use it in the clipboard pipeline, though, this would be partially against the  `module:image/imageinsert~ImageInsertConfig#type` documentation which states it works for new images only. And since we have no means to discover whether a pasted image is returning back to the same editor or comes from an external data source, I'd rather not use it here. This way copy/paste inside the editor will be at least more predictable.

Reply:
:+1: